base_tier_validation/models/tier_validation.py

```diff
@@ -0,0 +1,870 @@
+        """Not allowed fields to write when validation depending
+        on the given function."""
+        exceptions = records_exception_function()
+        not_allowed_fields = []
+        for val in vals:
+            if val not in exceptions:
+                not_allowed_fields.append(val)
+        if not not_allowed_fields:
+            return []
+
+        not_allowed_field_names, allowed_field_names = [], []
+        for fld_name, fld_data in self.fields_get(
+            not_allowed_fields + exceptions
+        ).items():
+            if fld_name in not_allowed_fields:
+                not_allowed_field_names.append(fld_data["string"])
+            else:
+                allowed_field_names.append(fld_data["string"])
+        return allowed_field_names, not_allowed_field_names
+
+    def write(self, vals):
+        self._tier_validation_check_state_on_write(vals)
+        self._tier_validation_check_write_allowed(vals)
+        self._tier_validation_check_write_remove_reviews(vals)
+        return super().write(vals)
+
+    def _write(self, vals):
+        if self._tier_validation_state_field_is_computed:
+            self._tier_validation_check_state_on_write(vals)
+            self._tier_validation_check_write_remove_reviews(vals)
+        return super()._write(vals)
+
+    def _write_multi(self, vals_list):
+        for rec, vals in zip(self, vals_list, strict=False):
+            if rec._tier_validation_state_field_is_computed:
+                rec._tier_validation_check_state_on_write(vals)
+                rec._tier_validation_check_write_remove_reviews(vals)
+        return super()._write_multi(vals_list)
+
+    def _tier_validation_get_current_state_value(self):
+        """Get the current value from the cache or the database.
+
+        If the field is set in a computed method, the value in the cache will
+        already be the updated value, so we need to revert to the raw data.
+        """
+        self.ensure_one()
+        if self._tier_validation_state_field_is_computed and isinstance(self.id, int):
+            self.env.cr.execute(
+                "select %(field)s from %(table)s where id = %(res_id)s",
+                {
+                    "field": AsIs(self._state_field),
+                    "table": AsIs(self._table),
+                    "res_id": self.id,
+                },
+            )
+            rows = self.env.cr.fetchall()
+            if rows:
+                return rows[0][0]
+        return self[self._state_field]
+
+    def _tier_validation_check_state_on_write(self, vals):
+        for rec in self:
+            if rec._check_state_conditions(vals):
+                if rec.need_validation:
+                    # try to validate operation
+                    reviews = rec.request_validation()
+                    rec._validate_tier(reviews)
+                    if not self._calc_reviews_validated(reviews):
+                        raise ValidationError(
+                            _(
```

Comment:
```suggestion
                            self.env._(
```
detail in:  https://github.com/odoo/odoo/commit/b794f0f332f473deb2c04eba60baf4761db3b508